README.md
![Actions Status](https://github.com/NicolaeIotu/typelike/workflows/tests/badge.svg)
![Actions Status](https://github.com/NicolaeIotu/typelike/workflows/max%20coverage/badge.svg)
![Actions Status](https://github.com/NicolaeIotu/typelike/workflows/standard%20code%20style/badge.svg)

# typelike

With **typelike** you can reliably determine if an object resembles other template object(s) used as reference.<br>
The comparison is done using the keys and the type of data where applicable.<br>
In practice using **typelike** is a breeze. You just get a trusted object and have **typelike** compare with a target
 object. That's all. Plain Javascript, no aliens, lightweight, zero dependencies, fast, easy to verify and use.

* [Examples](#examples)
  * [typelike in Node.js](#typelike-in-nodejs)
  * [typelike in browser](#typelike-in-browser)
* [How to compare with typelike](#how-to-compare-with-typelike)
* [Using multiple templates with typelike](#using-multiple-templates-with-typelike)
* [typelike Settings](#typelike-settings)
* [Others](#others)

## Examples
### typelike in Node.js
**Example 1:**
```
const { typelike } = require('typelike')

const testObject = {
  lvl1: { lvl2: [1, 2, 3], sm: 'type ... like' },
  arr: [[1, 'xyz'], 'abcdef']
}

const templateObject = {
  lvl1: { lvl2: [3, 4, 212], sm: '' },
  arr: [[44, ''], '']
}

console.log(typelike(testObject, templateObject)) // true
```
**Example 2:**
```
const { typelike } = require('typelike')

function contentGenerator () {
  // ...
  return {
    lvl1: {
      lvl2: {
        lvl3: {
          lvl4: {
            item1: [1, 2, 3],
            item2: true
          }
        }
      }
    },
    arr: [[1, 'xyz'], 'abcdef'],
    basic: 'testtesttest'
  }
}

const testObject = contentGenerator()

const templateObject = {
  lvl1: {
    lvl2: {
      lvl3: {
        lvl4: {
          item1: [44, 66, 88],
          item2: false
        }
      }
    }
  },
  arr: [[45, 'sample'], String('string')],
  basic: 'testtesttest'
}

console.log(typelike(testObject, templateObject)) // true
```

### typelike in browser
```
...
<head>
    ...
    <script src="./typelike.bundle.js"></script>
    ...
...
    <script>

      const typelike = typelikeBundled.typelike
    
      const testObject = {
        lvl1: { lvl2: [1, 2, 3], sm: 'type ... like' },
        arr: [[1, 'xyz'], 'abcdef']
      }
    
      const templateObject = {
        lvl1: { lvl2: [3, 4, 212], sm: '' },
        arr: [[44, ''], '']
      }
    
      console.log(typelike(testObject, templateObject)) // true
    
    </script>
...
```

## How to compare with typelike
**typelike** iterates as deeply as possible any arrays, objects, maps and sets, while taking into account 
[typelike Settings](#typelike-settings).
For type checking **typelike** uses <a href="https://github.com/NicolaeIotu/xtypeof" title="xtypeof" target="_blank">**xtypeof**</a> 
which is included as part of the application.

In order to pass **typelike** tests the following logic applies:
* for **arrays, objects and maps** tests pass if:
  - keys are identical for both subject object and template object and ...
  - the type of data corresponding to a key is identical for both subject object and template object.
  - if the data corresponding to a key is iterable as understood by **typelike** (array, object, map, set) the checks
    continue at the next deeper level
* for **sets** tests pass if:
  - the type of data corresponding to an entry is identical for both subject set and template set.
  - if the data corresponding to an entry is iterable as understood by **typelike** (array, object, map, set) the checks 
    continue at the next deeper level
* for any other type tests pass if:
  - the type of subject data is identical for both subject object and template object

## Using multiple templates with typelike
Multiple templates can be used with every call to `typelike`. If the subject object (which should be the first
 parameter specified) matches any of them then `typelike` will return true.
 **Example:**
```
const { typelike } = require('typelike')

const testObject = {
  lvl1: { lvl2: [1, 2, 3], sm: 'type ... like' },
  arr: [[1, 'xyz'], 'abcdef']
}

const template1 = {
  lvl1: {
    lvl2: {
      lvl3: {
        lvl4: {
          item1: [1, 2, 3],
          item2: true
        }
      }
    }
  },
  arr: [[1, 'xyz'], 'abcdef'],
  basic: 'testtesttest'
}
const template2 = {
  lvl1: { sm: 'type ... like' },
  arr: [[1], 'abcdef']
}
const template3 = {
  lvl1: { lvl2: [3, 4, 212], sm: '' },
  arr: [[44, ''], '']
}

console.log(typelike(testObject, template1, template2)) // false
console.log(typelike(testObject, template1, template2, template3)) // true
```

## typelike Settings
**typelike** behavior can be changed with a couple of settings. In order to run with altered the settings, a special
 function `typelikeCustom` is exported. The settings are valid for a single call to `typelikeCustom`. Subsequent
  calls to `typelike` will use the default settings as listed below.
  
`typelikeCustom` behaves in the same way as `typelike`, but expects that the last parameter passed to be a special
 setting object which takes the following default format/values:
```
{
  maxDepth: 0,
  properties: {
    allowMissing: false,
    allowNull: false
  }
}
```
* **maxDepth** indicates the maximum depth allowed for iterations. Defaults to `0` (unlimited depth levels)
* **properties.allowMissing** takes a boolean value and indicates that the keys/properties are mandatory (`false`), or 
  not mandatory and can miss from arrays, objects, maps and sets (`true`). Defaults to `false` (all keys/properties are 
  mandatory)
* **properties.allowNull** takes a boolean value and indicates that **null** value keys/properties of a target object,
  match **any** corresponding types in the template object(s) (`true`). Defaults to `false` which means that **null** 
  value keys/properties of a target object match corresponding **null** value keys/properties of template object(s).

**Example:**
```
const { typelikeCustom } = require('typelike')

const testObject = {
  lvl1: { sm: 'type ... like' },
  arr: [[1], 'abcdef']
}

const templateObject = {
  lvl1: { lvl2: [3, 4, 212], sm: '' },
  arr: [[44, ''], '']
}

const settings = {
  maxDepth: 3,
  properties: {
    allowMissing: true
  }
}

console.log(typelikeCustom(testObject, templateObject, settings)) // true
```

## Others
**typelike** should fit most target purposes. For some cases remember you can use as many templates as needed. 
 
Additional settings are planned for the next releases. For suggestions and issues please contact the author.


**typelike** is &copy; Copyright 2020-2021 Nicolae Iotu, nicolae.g.iotu@gmail.com


CHANGELOG.md
# typelike Changelog

## v0.4.0 Backend improvements to documentation, release and distribution

## v0.3.9 Added setting option (allowNull) to match null values against all types

## v0.3.8 Backend and documentation improvements

## v0.3.7 Backend improvements

## v0.3.6 Improved efficiency
Improved maps and sets logic by eliminating double checks of properties.allowMissing

## v0.2.4 Correct error eliminate babel and dist/typelike.js

## v0.1.3 Add package-lock.json and Correct error cases
Correct some errors with the handling of basic types
Add tests to support basic types
Add package-lock.json



## v0.1.0 Initial release


lib/typelike.js
'use strict'

/* typelike is Copyright (C) 2020 Nicolae Iotu, nicolae.g.iotu@gmail.com
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
limitations under the License.

Author Nicolae Iotu, nicolae.g.iotu@gmail.com */

const hasOwnProperty = Object.prototype.hasOwnProperty

// xtypeof module
function xtypeof (o) {
  return Object.prototype.toString.call(o).slice(8, -1).toLowerCase()
}

function iterationType (o) {
  switch (xtypeof(o)) {
    case 'object':
      return 4
    case 'array':
      return 3
    case 'map':
      return 2
    case 'set':
      return 1
    default:
      return 0
  }
}

const templateSettings = {
  maxDepth: 0,
  properties: {
    allowMissing: false,
    allowNull: false
  }
}

let typelikeSettings = templateSettings
let rtDepth = 0

const resetOptions = function () {
  typelikeSettings = templateSettings
  rtDepth = 0
}

function compare (o, t) {
  const ito = iterationType(o)
  const itt = iterationType(t)

  if (ito !== itt) {
    return false
  }

  if (ito !== 0) {
    // depth checks
    if (typelikeSettings.maxDepth !== 0) {
      rtDepth++
      if (rtDepth > typelikeSettings.maxDepth) {
        return true
      }
    }

    if (ito >= 3) {
      const oSize = Object.keys(o).length
      const tSize = Object.keys(t).length
      if (!typelikeSettings.properties.allowMissing) {
        if (oSize === 0) {
          return oSize === tSize
        } else if (oSize !== tSize) {
          return false
        }
      }

      for (const k in o) {
        if (hasOwnProperty.call(t, k + '')) {
          if (!compare(o[k], t[k])) {
            return false
          }
        } else if (!typelikeSettings.properties.allowMissing) {
          // allowMissing logic
          return false
        }
      }
    } else {
      // Map and Set logic
      const oSize = o.size
      const tSize = t.size

      if (!typelikeSettings.properties.allowMissing) {
        if (oSize === 0) {
          return oSize === tSize
        } else if (oSize !== tSize) {
          return false
        }
      }

      const oKeysIterator = o.keys()
      const tKeysIterator = t.keys()
      if (ito === 1) {
        // sets
        for (let i = 0; i < oSize; i++) {
          if (!compare(oKeysIterator.next().value, tKeysIterator.next().value)) {
            return false
          }
        }
      } else {
        // maps
        const oValuesIterator = o.values()
        const tValuesIterator = t.values()
        for (let i = 0; i < oSize; i++) {
          if (oKeysIterator.next().value === tKeysIterator.next().value) {
            if (!compare(oValuesIterator.next().value, tValuesIterator.next().value)) {
              console.log(111)
              return false
            }
          } else {
            return false
          }
        }
      }
    }

    // depth logic
    if (typelikeSettings.maxDepth !== 0) {
      rtDepth--
    }

    return true
  } else {
    return (xtypeof(o) === xtypeof(t) || (typelikeSettings.properties.allowNull && xtypeof(o) === 'null'))
  }
}

/**
 * With **typelike** you can reliably determine if an object resembles other template object(s) used as reference.<br>
 * The comparison is done using the keys and the type of data where applicable.<br>
 * **typelike** iterates as deeply as possible any arrays, objects, maps and sets, while taking into account the
 * standard settings:
 * ```
 * {
 *     maxDepth: 0,
 *     properties: {
 *         allowMissing: false,
 *         allowNull: false
 *     }
 * }
 * ```
 * For type checking **typelike** uses
 * <a href="https://github.com/NicolaeIotu/xtypeof" title="xtypeof" target="_blank">**xtypeof**</a> which is
 * included as part of the application.<br>
 * @param obj {object} A test object to match against the template(s).
 * @param templates {object} Templates are trusted objects. The test object `obj` must match in depth the types of
 * these templates.
 * @returns {boolean} **true** if the target object `obj` matches any of the templates provided, or **false** otherwise
 * @version 0.3.8
 * @license Apache-2.0
 * @author Nicolae Iotu <nicolae.g.iotu@gmail.com>
 * @exports typelike
 * @example
 * const { typelike } = require('typelike')
 * const testObject = {
 *     lvl1: { lvl2: [1, 2, 3], sm: 'type ... like' },
 *     arr: [[1, 'xyz'], 'abcdef']
 * }
 * const templateObject = {
 *     lvl1: { lvl2: [3, 4, 212], sm: '' },
 *     arr: [[44, ''], '']
 * }
 * console.log(typelike(testObject, templateObject)) // true
 * @example
 * // Using multiple templates with typelike
 * const { typelike } = require('typelike')
 * const testObject = {
 *     lvl1: { lvl2: [1, 2, 3], sm: 'type ... like' },
 *     arr: [[1, 'xyz'], 'abcdef']
 * }
 * const template1 = {
 *     lvl1: {
 *         lvl2: {
 *             lvl3: {
 *                 lvl4: {
 *                     item1: [1, 2, 3],
 *                     item2: true
 *                 }
 *             }
 *         }
 *     },
 *     arr: [[1, 'xyz'], 'abcdef'],
 *     basic: 'testtesttest'
 * }
 * const template2 = {
 *     lvl1: { sm: 'type ... like' },
 *     arr: [[1], 'abcdef']
 * }
 * const template3 = {
 *     lvl1: { lvl2: [3, 4, 212], sm: '' },
 *     arr: [[44, ''], '']
 * }
 * console.log(typelike(testObject, template1, template2)) // false
 * console.log(typelike(testObject, template1, template2, template3)) // true
 */
module.exports.typelike = function (obj, templates) {
  const lastIndex = arguments.length - 1
  if (lastIndex < 1) {
    throw new Error('At least two parameters required.')
  }

  let result = false
  for (let i = 1; i <= lastIndex; i++) {
    result = result || compare(obj, arguments[i])
  }

  // important! reset options
  resetOptions()

  return result
}

/**
 * **typelikeCustom** behaves in the same way as {@link module:typelike|typelike}, but expects that the last parameter
 * passed to
 * be a special settings object which is mandatory and has the following default format for standard
 * {@link module:typelike|typelike} calls:
 * ```
 * {
 *     maxDepth: 0,
 *     properties: {
 *         allowMissing: false,
 *         allowNull: false
 *     }
 * }
 *```
 *
 * The comparison is done using the keys and the type of data where applicable.<br>
 * **typelikeCustom** iterates as deeply as possible any arrays, objects, maps and sets, while taking into account the
 * settings.
 * For type checking **typelikeCustom** uses
 * <a href="https://github.com/NicolaeIotu/xtypeof" title="xtypeof" target="_blank">**xtypeof**</a> which is
 * included as part of the application.<br>
 * @param obj {object} A test object to match against the template(s).
 * @param templates {object} Templates are trusted objects. The test object `obj` must match in depth the types of
 * these templates.
 * @param settings {{maxDepth: number, properties: {allowMissing: boolean}}} The settings object to be used for this
 * call only.
 * @returns {boolean} **true** if the target object `obj` matches any of the templates provided, or **false** otherwise
 * @version 0.3.8
 * @license Apache-2.0
 * @author Nicolae Iotu <nicolae.g.iotu@gmail.com>
 * @exports typelikeCustom
 * @example
 * const { typelikeCustom } = require('typelike')
 * const testObject = {
 *     lvl1: { sm: 'type ... like' },
 *     arr: [[1], 'abcdef']
 * }
 * const templateObject = {
 *     lvl1: { lvl2: [3, 4, 212], sm: '' },
 *     arr: [[44, ''], '']
 * }
 * const settings = {
 *     maxDepth: 3,
 *     properties: {
 *         allowMissing: true,
 *         allowNull: true
 *     }
 * }
 * console.log(typelikeCustom(testObject, templateObject, settings)) // true
 */
module.exports.typelikeCustom = function (obj, templates, settings) {
  const lastIndex = arguments.length - 1
  if (lastIndex < 2) {
    throw new Error('At least three parameters required.')
  }
  typelikeSettings = arguments[lastIndex]
  return module.exports.typelike.apply(null, Array.from(arguments).filter(
    function (e, i) {
      return (i !== lastIndex)
    }))
}
